src/core/server/ui_settings/saved_objects/transforms.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { SavedObjectsErrorHelpers } from '../../saved_objects';
+import type { SavedObjectsClientContract } from '../../types';
+import type { UpgradeableConfigAttributes } from '../create_or_upgrade_saved_config';
+
+interface Params {
+  savedObjectsClient: SavedObjectsClientContract;
+  configAttributes:
+    | Pick<UpgradeableConfigAttributes, 'defaultIndex' | 'isDefaultIndexMigrated'>
+    | undefined;
+}
+
+/** The resulting parameters that should be used when upgrading the config object. */
+interface ReturnType {
+  isDefaultIndexMigrated: boolean;
+  defaultIndex?: string;
+}
+
+export type TransformConfigFn = (params: Params) => Promise<ReturnType>;
+
+/**
+ * Any transforms that should be applied during `createOrUpgradeSavedConfig` need to be included in this array.
+ */
+export const transforms: TransformConfigFn[] = [transformDefaultIndex];
+
+/**
+ * This optionally transforms the `defaultIndex` attribute of a config saved object. The `defaultIndex` attribute points to a data view ID,
+ * but those saved object IDs were regenerated in the 8.0 upgrade. That resulted in a bug where the `defaultIndex` would be broken in custom
+ * spaces.
+ *
+ * We are fixing this bug after the fact in 8.3, and we can't retroactively change a saved object that's already been migrated, so we use
+ * this transformation instead to ensure that the `defaultIndex` attribute is not broken.
+ *
+ * Note that what used to be called "index patterns" prior to 8.0 have been renamed to "data views", but the object type cannot be changed,
+ * so that type remains `index-pattern`.
+ *
+ * Note also that this function is only exported for unit testing. It is also included in the `transforms` export above, which is how it is
+ * applied during `createOrUpgradeSavedConfig`.
+ */
+export async function transformDefaultIndex(params: Params): Promise<ReturnType> {
+  const { savedObjectsClient, configAttributes } = params;
+  if (!configAttributes || !configAttributes.defaultIndex) {
```

Comment:
Don't we also want to exit if `isDefaultIndexMigrated` is already true, given these transformations will be executed on each version upgrade?